### ISSUE TITLE
Background and Design Draft

### ISSUE THREAD

**Issue by b0nes164** [b0nes164/Decoupled-Fallback-Paper#3]
Background and Design sections are ready for review.

Also, I had a question about something from the Merrill and Grimshaw [paper](https://libraopen.lib.virginia.edu/downloads/6t053g00z).

In a section comparing the CUDPP intra-workgroup strategy Merrill writes: `In addition, the use of work-inefficient SIMD Kogge-Stone "warp-scans" for processing phases other than the top-phase of a cooperative threadblock cycle results in a much higher dynamic instruction count than necessary`. 
 
Again referring to the instruction counts of CUDPP scan kernel he later writes: `When only considering scan kernels, the comparative efficiency of our scan designs is still significant, but not as dramatic. The merrill_tree, merrill_srts, and merrill_linear scan kernels execute 18%, 44%, and 60% fewer instructions than the CUDPP scan kernel, respectively. On paper, the CUDPP scan kernel instruction overhead exceeds the cap of 17.9 instructions-per-element, implying that no amount of data-movement tuning will allow it to consume the GTX-285’s theoretical maximum of 159 GiB/sec of device memory bandwidth. However, the accelerator appears to provide a physically achievable maximum of 138 GiB/sec for in-
out kernels, resulting in a more realistic overhead cap of 20.6 instructions-per-element. This would give our particular hardware an extremely small margin of error (~6%) in which to optimally schedule a hypothetical bandwidth-tuned CUDPP scan computation, making it unlikely that it could be brought up to speed with our implementations on our test platform.`

This was particularly concerning because, by coincidence, my current scan formulation is extremely similar to CUDPP, the only difference is that [I rake my subgroups](https://github.com/b0nes164/GPUPrefixSums/blob/98d93a4e9ed2f3c8353119515bf9be90a2e137ad/GPUPrefixSumsWebGPUapis/SharedShaders/csdldf.wgsl#L105-L108) before posting the subgroup reduction to shared memory. Based on Merrill's commentary, I was worried that the current scan-then-propagate strategy would be inferior to reduce-then-scan, and that the [use of subgroup-scans](https://github.com/b0nes164/GPUPrefixSums/blob/98d93a4e9ed2f3c8353119515bf9be90a2e137ad/GPUPrefixSumsWebGPUapis/SharedShaders/csdldf.wgsl#L118-L149) for the workgroup-wide scan would be inferior to a simpler non-subgroup Brent-Kung scan.

However after testing [reduce-then-scan ](https://github.com/b0nes164/GPUPrefixSums/blob/783d9f6d759c00fff418fb91d19285236fb1d752/GPUPrefixSumsWebGPUapis/SharedShaders/csdldf.wgsl#L251-L277), Merrill's [Brent-Kung scan with bank conflict avoidance](https://github.com/b0nes164/GPUPrefixSums/blob/3b595d8407cad287e223bc257829bf9f077b952f/GPUPrefixSumsWebGPUapis/SharedShaders/csdldf.wgsl#L117-L152), and [MatrixScan's raking with bank conflict avoidance](https://github.com/b0nes164/GPUPrefixSums/blob/ed45aca196c3c0dd08a9875b8cf0cc4a7d70feb6/GPUPrefixSumsWebGPUapis/SharedShaders/csdldf.wgsl#L98-L119), each was slower than my current CUDPP style scan across all the hardware I tested on (2080 Super,  Mali G78, Intel HD620).

I can understand Yan et al.'s argument for reduce-then-scan in StreamScan---they want the workgroup reductions available as quickly as possible due to the propagation latency bottlenecking---but I don't understand Merrill's argument for reduced instruction counts. For example, he advocates for reduce-then-scan because it requires less instructions, but the data on my hardware suggests the redundant computation makes the strategy slower than scan-then-propagate.

Are instruction counts 1:1 with compute time? Is counting instructions still a relevant form of analysis? As a rule of thumb, do contemporary GPUs typically run into instruction count bottlenecking?

**Comment by jowens** [comment]
So there's two possibilities, right? Let's assume your data is correct, because of course it is.
- Merrill was wrong.
- Merrill was right, and the hardware from 15 years ago is simply different than today's hardware.

Either seems reasonable. It's hard to make a conclusion because you don't have 2009 GPUs, but we can discuss it. If I'm understanding this right, the tradeoff is between two forms of compute (less redundant computation vs. reduction of critical path)? This is well worth discussing. (I love this article btw: https://dl.acm.org/doi/10.1145/3213770 ).

> Are instruction counts 1:1 with compute time? Is counting instructions still a relevant form of analysis? As a rule of thumb, do contemporary GPUs typically run into instruction count bottlenecking?

I'm just guessing here, but: If you truly are bottlenecked by compute, then counting instructions is a reasonable approach IMHO. Some instructions cost more than others. It gets hard when there are barriers. Memory latencies obviously make it really hard. My experience is that not many kernels are bottlenecked by compute, but there are definitely some. My students would tell me the NVIDIA perf tools are pretty good. I definitely have students who have a LOT more experience with this than I do.

Aside: One thing Merrill did that I really liked was to show that he had basically reached the speed of light (with the 3n memory cost) and future improvements didn't matter. He was right, and it was a good way to characterize it, so that future research didn't focus on optimizing something that didn't matter.

Separate aside: Do you have library access to literature? Perhaps through UCLA? If you don't, lemme know, I'll hook you up.

**Comment by jowens** [comment]
As I'm writing this, I can see that you might be thinking "why is John making me do all this work, why doesn't he just write it himself rather than telling me what to do". I can write sections and you can assign me things if that makes sense to you. I'm doing what I'm doing for two reasons: (1) you are a good writer and I am confident that you writing these sections will yield great text; (2) I hope that the comments I make will help you be a better _academic_ writer and present the ideas / flow in a way that will make this paper as impactful as possible, and you listening to my comments and incorporating them into your writing is the way I think is most effective at making you think in the right way to present an academic-paper idea.

**Comment by b0nes164** [comment]
> As I'm writing this, I can see that you might be thinking "why is John making me do all this work, why doesn't he just write it himself rather than telling me what to do"

Absolutely no worries. As I told Raph before we agreed to do the paper, I'm perfectly happy to do all the leg work. Moreover, writing the paper myself has deepened my understanding of the material---understanding that I would not have gotten if you wrote it.

> Do you have library access to literature? Perhaps through UCLA? If you don't, lemme know, I'll hook you up.

I have access, but thank you.

> I love this article btw: https://dl.acm.org/doi/10.1145/3213770 

I'm almost certain I've heard an abridged version of this in a Charles Leiserson lecture, but I have definitely encountered the issues Ousterhout mentions. 

> So there's two possibilities, right? Let's assume your data is correct, because of course it is.

Merrill's work is typically excellent, which is why I'm inclined to believe that, like you said, Merrill was correct but current GPUs throw his calculations out of wack. Of course my implementations could also have been sub-optimal, but I think that points towards a deeper issue---profiling. I'm pretty familiar with Nsight Compute for profiling in CUDA (it's very useful for sanity checking performance results as Ousterhout mentions), but I'm not sure if similar tools exists for *all* backends of the WebGPU spec (D3D12 has PIX, not sure about Metal or Vulkan). So if we did want to probe this more, I'm not sure how we would go about it rigorously. 

And beyond that I'm not sure if such rigorous profiling across all backends would be a good use of time: once we show that we're at speed-of-light, any improvements beyond that are just power saving.

**Comment by jowens** [comment]
> So if we did want to probe this more, I'm not sure how we would go about it rigorously.

Let's not make the perfect the enemy of the good here. I'm not saying we need to do this now or at all, I think we'll have a better idea if we need to do this once we look at the results narrative, but "doing it rigorously on an NVIDIA architecture that has good perf tools" gets us meaningful results that we can discuss even if it's just on one architecture. Almost all the scan work you discuss was implemented on one and only one architecture, and that didn't stop it from being impactful at all.

[Edit: I have two recent PhD graduates at AMD Research who could offer good advice re: AMD hardware profiling if we go down that path.]

> Any improvements beyond that are just power saving

Or, as Duane has pointed out, can be used to do more compute (e.g., a transform-scan, "scan the square of each element of this input vector").

**Comment by jowens** [comment]
I am also happy to broker an conversation with Duane on this matter, I'm sure he'd be willing, but that's conditioned on you being really really ready to present it in a way that makes you look super smart, because (besides the benefits for this work), the primary purpose for that engagement is to enable you to gain Duane's respect. Which you will. Not urgent.

**Comment by b0nes164** [comment]
> I am also happy to broker an conversation with Duane on this matter

Absolutely, sometime after the paper is more settled. I just want to understand the line of analysis, and how to apply it for future work. Most papers I have read do not go that low level, and stick to metrics like throughput and %memory bandwidth utilization.

**Comment by jowens** [comment]
This is great stuff. Lots of inline comments checked in for you. But it's a good organization.

Typographically: footnotes go after punctuation, not before; cites are separated by a `~` from their citing text (`Foo and Bar~\cite{foo-bar}`; use `\@` before a period when an acronym ends a sentence (e.g.: `... on the GPU\@.`). Your math-mode typesetting is top-notch.
